src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java

```diff
@@ -702,7 +704,10 @@ public void serializeAsField(Object bean, JsonGenerator gen,
         }
         // then find serializer to use
         JsonSerializer<Object> ser = _serializer;
-        if (ser == null) {
+		if (ser == null || (value instanceof EnumSet && ser.getClass() != EnumSetSerializer.class)) {
+			// if value instance of EnumSet, use EnumSetSerializer for EnumSet instead of
+			// generic CollectionSerializer, since de-serialization uses EnumSetDeserializer,
+			// if property value is of type EnumSet
```

Comment:
I am sorry, but this is not a valid solution. The solution does not seem to fit overall and class design in any way. Is there simliar implementation in `BeanPropertyWriter`? Or any other pattern that you refered to?

Reply:
i agree that the fix is not fitting in overall design of this class, just added this workaround here to explain the problem better. I tried to set the serializer while initializing the instance of this class, but the runtime type information of the property value is not available at the time of initialization this class's object.